verilog/tools/ls/symbol-table-handler.cc

```diff
@@ -0,0 +1,249 @@
+    LOG(INFO) << "Adding include path:  " << incdir;
+    currproject->AddIncludePath(incdir);
+  }
+  // add files from file list to the project
+  for (auto &incfile : filelist.file_paths) {
+    auto incsource = currproject->OpenIncludedFile(incfile);
+    if (!incsource.ok()) {
+      LOG(WARNING) << "File included in " << projectpath
+                   << " not found:  " << incfile << ":  " << incsource.status();
+      continue;
+    }
+    LOG(INFO) << "Creating symbol table for:  " << incfile;
+    BuildSymbolTableFor(*incsource.value());
+  }
+}
+
+bool IsStringViewContained(absl::string_view origin, absl::string_view substr) {
+  const int from = std::distance(origin.begin(), substr.begin());
+  const int to = std::distance(origin.begin(), substr.end());
+  if (from < 0) return false;
+  if (to > static_cast<int>(origin.length())) return false;
+  return true;
+}
+
+const SymbolTableNode *ScanReferenceComponents(
+    const ReferenceComponentNode *ref, absl::string_view symbol) {
+  if (IsStringViewContained(ref->Value().identifier, symbol))
+    return ref->Value().resolved_symbol;
+  for (const auto &childref : ref->Children()) {
+    const SymbolTableNode *resolved =
+        ScanReferenceComponents(&childref, symbol);
+    if (resolved) return resolved;
+  }
+  return nullptr;
+}
+
+const SymbolTableNode *SymbolTableHandler::ScanSymbolTreeForDefinition(
+    const SymbolTableNode *context, absl::string_view symbol) {
+  if (!context) {
+    return nullptr;
+  }
+  // TODO (glatosinski): reduce searched scope by utilizing information from
+  // syntax tree?
+  for (const auto &ref : context->Value().local_references_to_bind) {
+    if (ref.Empty()) continue;
+    const SymbolTableNode *resolved =
+        ScanReferenceComponents(ref.components.get(), symbol);
+    if (resolved) return resolved;
+  }
+  for (const auto &child : context->Children()) {
+    auto res = ScanSymbolTreeForDefinition(&child.second, symbol);
+    if (res) {
+      return res;
+    }
+  }
+  return nullptr;
+}
+
+std::vector<verible::lsp::Location> SymbolTableHandler::FindDefinition(
+    const verible::lsp::DefinitionParams &params,
+    const verilog::BufferTrackerContainer &parsed_buffers) {
+  const absl::Time finddefinition_start = absl::Now();
+  if (files_dirty_) {
+    BuildProjectSymbolTable();
+  }
+  absl::string_view filepath = LSPUriToPath(params.textDocument.uri);
+  if (filepath.empty()) {
+    LOG(ERROR) << "Could not convert URI " << params.textDocument.uri
+               << " to filesystem path." << std::endl;
+    LOG(INFO) << "textDocument/definition processing time:  "
+              << absl::ToInt64Milliseconds(absl::Now() - finddefinition_start)
+              << "ms";
+    return {};
+  }
+  std::string relativepath = currproject->GetRelativePathToSource(filepath);
+  const verilog::ParsedBuffer *parsedbuffer =
+      parsed_buffers.FindBufferTrackerOrNull(params.textDocument.uri)
```

Comment:
`FindBufferTrackerOrNull()` can return nullptr.
Here, we're relying on the editor sending us a valid `params.textDocument.uri` that we have store. But if not, we'd crash here. So I'd apply defensive programming here to test for that early.

(anything where we depend on external information that could make us crash ... doublechecking is good).